### PR TITLE
DL-445 celery director dynamic task module loading

### PR DESCRIPTION
**Pull request by CodePint**
Implementing celery dynamic task module loading for director

This PR retains backwards functionality for the existing task loading whilst adding the ability
to dynamically load and run tasks using celery signatures.

By doing this we are freed of the limitation of storing tasks in workflow-engine `taskflow/tasks` and are able
to further separate api and task logic within workflow-engine tasks

refs:
https://docs.celeryproject.org/en/stable/userguide/configuration.html#example-configuration-file
https://docs.celeryproject.org/en/stable/reference/celery.html#celery.signature

ticket: https://linius.atlassian.net/browse/DL-445

### CODE REVIEW

File: director/settings.py [hunk 1]
```diff
@@ -64,9 +64,14 @@ def __init__(self, home_path=None, config_path=None):
             "result_backend": env.str(
                 "DIRECTOR_RESULT_BACKEND_URI", "redis://localhost:6379/1"
             ),
+            "imports": ["director.tasks"],
             "broker_transport_options": {"master_name": "director"},
         }
 
+        # Celery task import config
+        if any(env.list("DIRECTOR_CELERY_IMPORTS")):
+            self.CELERY_CONF['imports'] += env.list("DIRECTOR_CELERY_IMPORTS")
```
Comment (AlanDonohoe) [inline director/settings.py]: Inconsistent single vs double quotes.

File: tests/conftest.py [hunk 2]
```diff
@@ -43,8 +43,9 @@ def json(self):
         return _remove_keys(self.get_json(), self._KEYS_TO_REMOVE)
 
 
-@pytest.fixture(scope="module")
+@pytest.fixture(scope="function")
 def app_module():
+    os.environ['DIRECTOR_CELERY_IMPORTS'] = 'tests.workflows.tasks'
```
Comment (AlanDonohoe) [inline tests/conftest.py:48]: Inconsistent single vs double quotes.

File: director/settings.py (hunk 1 above)
Comment (CodePint) [reply director/settings.py]: good catch...damn black requirement.

File: tests/conftest.py (hunk 2 above)
Comment (CodePint) [reply tests/conftest.py:48]: black again...

**Comment by AlanDonohoe** [comment]
Is this extra complexity warranted in that we can organise tasks in structured way just using this:

```
foo/__init__.py
               tasks.py
               other_code.py

            bar/__init__.py
                tasks.py
                other_code.py

            baz/__init__.py
                models.py

```

ref: 
https://stackoverflow.com/questions/17702578/how-to-structure-celery-tasks

And the autodiscover option doesn't seem necessary if the various tasks.py files are placed in the top level tasks directory.

I am working in this area now. So I have in WFE:

taskflow/workflows.yml

```yaml
taskflow.ocr:
  tasks:
    - OCR_DETECT_TEXT
    - OCR_PARSE_DATA
    - OCR_CREATE_LVS_CLIPS
```

And in:
taskflow/tasks/ocr/tasks.py

```python
from director import task

from taskflow.tasks.ocr.detect_text import DetectText


@task(name="OCR_DETECT_TEXT")
def detect_text(*args, **kwargs):
    video_url = kwargs["payload"]["video_url"]
    detect_text = DetectText(video_url)
    detect_text.run()
```


So the tasks.py file just registers and defines the very simple task and then calls the more complex implementation that can live anywhere in the app. 

Which means we have the tasks organised in a sensible location 
tasks/ocr/tasks.py
tasks/transcribe/tasks.py
etc


Rather than having them scattered throughout the code base.

But they are more like thin controllers that can then access more complex classes, modules etc which are saved somewhere else sensible in the codebase.

**Review by CygnusAlpha — APPROVED** [review]

**Review by acr26 — CHANGES_REQUESTED** [review]
PR into base repo

**Pull request closed without merging**